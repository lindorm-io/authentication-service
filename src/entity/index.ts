export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
