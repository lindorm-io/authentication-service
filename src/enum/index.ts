export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
