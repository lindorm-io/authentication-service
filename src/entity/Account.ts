import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
