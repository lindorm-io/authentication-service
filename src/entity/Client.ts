import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { ClientEvent } from "../enum";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  emailAuthorizationUri?: string;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _emailAuthorizationUri: string;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._emailAuthorizationUri = options?.emailAuthorizationUri || null;
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get emailAuthorizationUri(): string {
    return this._emailAuthorizationUri;
  }
  public set emailAuthorizationUri(uri: string) {
    this._emailAuthorizationUri = uri;
    this.addEvent(ClientEvent.EMAIL_AUTHORIZATION_URI_CHANGED, { uri: this._emailAuthorizationUri });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new Error("Client has already been created");
    }
    this.addEvent(ClientEvent.CREATED, {
      name: this._name,
      emailAuthorizationUri: this._emailAuthorizationUri,
      approved: this._approved,
      description: this._description,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
