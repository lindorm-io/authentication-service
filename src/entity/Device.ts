import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { DeviceEvent } from "../enum";

export interface IDevicePIN {
  signature: string;
  updated: Date;
}

export interface IDevice extends IEntity {
  accountId: string;
  name: string;
  pin: IDevicePIN;
  publicKey: string;
  secret: string;
}

export interface IDeviceOptions extends IEntityBaseOptions {
  accountId: string;
  name?: string;
  pin?: IDevicePIN;
  publicKey: string;
  secret?: string;
}

export class Device extends EntityBase implements IDevice {
  private _accountId: string;
  private _name: string;
  private _pin: IDevicePIN;
  private _publicKey: string;
  private _secret: string;

  constructor(options: IDeviceOptions) {
    super(options);
    this._accountId = options.accountId;
    this._name = options.name || null;
    this._pin = {
      signature: options.pin?.signature || null,
      updated: options.pin?.updated || null,
    };
    this._publicKey = options.publicKey;
    this._secret = options.secret || null;
  }

  public get accountId(): string {
    return this._accountId;
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(DeviceEvent.NAME_CHANGED, { name: this._name });
  }

  public get pin(): IDevicePIN {
    return this._pin;
  }
  public set pin(pin: IDevicePIN) {
    this._pin = pin;
    this.addEvent(DeviceEvent.PIN_CHANGED, { pin: this._pin });
  }

  public get publicKey(): string {
    return this._publicKey;
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(DeviceEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DeviceEvent.CREATED) continue;
      throw new Error("Device has already been created");
    }
    this.addEvent(DeviceEvent.CREATED, {
      accountId: this._accountId,
      name: this._name,
      pin: this._pin,
      publicKey: this._publicKey,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
