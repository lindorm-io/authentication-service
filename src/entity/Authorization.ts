import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { GrantType } from "../enum";
import { Scope } from "@lindorm-io/jwt";

export interface IAuthorization extends IEntity {
  challengeId: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export interface IAuthorizationOptions extends IEntityBaseOptions {
  challengeId?: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId?: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export class Authorization extends EntityBase implements IAuthorization {
  readonly challengeId: string;
  readonly clientId: string;
  readonly codeChallenge: string;
  readonly codeMethod: string;
  readonly deviceId: string;
  readonly email: string;
  readonly expires: Date;
  readonly grantType: GrantType;
  readonly otpCode: string;
  readonly redirectUri: string;
  readonly responseType: string;
  readonly scope: Array<Scope>;

  constructor(options: IAuthorizationOptions) {
    super(options);

    this.challengeId = options.challengeId || null;
    this.clientId = options.clientId;
    this.codeChallenge = options.codeChallenge;
    this.codeMethod = options.codeMethod;
    this.deviceId = options.deviceId || null;
    this.email = options.email;
    this.expires = options.expires;
    this.grantType = options.grantType;
    this.otpCode = options.otpCode || null;
    this.redirectUri = options.redirectUri;
    this.responseType = options.responseType;
    this.scope = options.scope;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
