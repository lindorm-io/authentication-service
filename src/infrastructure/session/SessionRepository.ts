import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { ISession, Session } from "../../entity";
import { MongoCollection } from "../../enum";
import { indices } from "./indices";
import { schema } from "./schema";

export interface ISessionFilter {
  id?: string;
  accountId?: string;
  authenticated?: boolean;
  clientId?: string;
  deviceId?: string;
  expires?: Date;
  grantType?: string;
  refreshId?: string;
}

export interface ISessionRepository extends IRepository<Session> {
  create(entity: Session): Promise<Session>;
  update(entity: Session): Promise<Session>;
  find(filter: ISessionFilter): Promise<Session>;
  findMany(filter: ISessionFilter): Promise<Array<Session>>;
  findOrCreate(filter: ISessionFilter): Promise<Session>;
  remove(entity: Session): Promise<void>;
}

export class SessionRepository extends RepositoryBase<Session> implements ISessionRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.SESSION,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: ISession): Session {
    return new Session(data);
  }

  protected getEntityJSON(entity: Session): ISession {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      accountId: entity.accountId,
      agent: entity.agent,
      authenticated: entity.authenticated,
      authorization: entity.authorization,
      clientId: entity.clientId,
      deviceId: entity.deviceId,
      expires: entity.expires,
      grantType: entity.grantType,
      refreshId: entity.refreshId,
      scope: entity.scope,
    };
  }

  public async create(entity: Session): Promise<Session> {
    return super.create(entity);
  }

  public async find(filter: ISessionFilter): Promise<Session> {
    return super.find(filter);
  }

  public async findMany(filter: ISessionFilter): Promise<Array<Session>> {
    return super.findMany(filter);
  }

  public async remove(entity: Session): Promise<void> {
    await super.remove(entity);
  }
}
