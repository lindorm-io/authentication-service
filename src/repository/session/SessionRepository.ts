import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { ISession, Session } from "../../entity";
import { MongoCollection } from "../../enum";
import { indices } from "./SessionRepository.indices";
import { schema } from "./SessionRepository.schema";

export interface ISessionFilter {
  id?: string;
  accountId?: string;
  authenticated?: boolean;
  clientId?: string;
  deviceId?: string;
  expires?: Date;
  grantType?: string;
  refreshId?: string;
}

export interface ISessionRepository extends IRepository<Session> {
  create(session: Session): Promise<Session>;
  update(session: Session): Promise<Session>;
  find(filter: ISessionFilter): Promise<Session>;
  findMany(filter: ISessionFilter): Promise<Array<Session>>;
  remove(session: Session): Promise<void>;
}

export class SessionRepository extends RepositoryBase<Session> implements ISessionRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.SESSION,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: ISession): Session {
    return new Session(data);
  }

  protected getEntityJSON(session: Session): ISession {
    return {
      id: session.id,
      version: session.version,
      created: session.created,
      updated: session.updated,
      events: session.events,

      accountId: session.accountId,
      agent: session.agent,
      authenticated: session.authenticated,
      authorization: session.authorization,
      clientId: session.clientId,
      deviceId: session.deviceId,
      expires: session.expires,
      grantType: session.grantType,
      refreshId: session.refreshId,
      scope: session.scope,
    };
  }

  public async create(session: Session): Promise<Session> {
    return super.create(session);
  }

  public async find(filter: ISessionFilter): Promise<Session> {
    return super.find(filter);
  }

  public async findMany(filter: ISessionFilter): Promise<Array<Session>> {
    return super.findMany(filter);
  }

  public async remove(session: Session): Promise<void> {
    await super.remove(session);
  }
}
