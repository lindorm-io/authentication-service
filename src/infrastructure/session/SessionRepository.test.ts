import MockDate from "mockdate";
import { GrantType, ResponseType } from "../../enum";
import { Scope } from "@lindorm-io/jwt";
import { RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { Session } from "../../entity";
import { SessionRepository } from "./SessionRepository";
import { getTestRepository, resetStore } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("SessionRepository", () => {
  let repository: SessionRepository;
  let session: Session;

  beforeEach(async () => {
    ({ session: repository } = await getTestRepository());

    session = new Session({
      id: "be3a62d1-24a0-401c-96dd-3aff95356811",

      accountId: "be3a62d1-24a0-401c-96dd-3aff95356811",
      agent: {
        browser: "browser",
        geoIp: { geoIp: "geoIp" },
        os: "os",
        platform: "platform",
        source: "source",
        version: "version",
      },
      authenticated: false,
      authorization: {
        codeChallenge: "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
        codeMethod: "sha256",
        deviceChallenge: "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
        email: "email@lindorm.io",
        id: "be3a62d1-24a0-401c-96dd-3aff95356811",
        otpCode: "otpCode",
        redirectUri: "https://lindorm.io/",
        responseType: [ResponseType.REFRESH, ResponseType.ACCESS, ResponseType.IDENTITY].join(" "),
      },
      clientId: "be3a62d1-24a0-401c-96dd-3aff95356811",
      deviceId: "be3a62d1-24a0-401c-96dd-3aff95356811",
      expires: new Date(),
      grantType: GrantType.EMAIL_OTP,
      refreshId: "be3a62d1-24a0-401c-96dd-3aff95356811",
      scope: [Scope.DEFAULT],
    });
  });

  afterEach(resetStore);

  test("should create", async () => {
    await expect(repository.create(session)).resolves.toMatchSnapshot();
  });

  test("should update", async () => {
    await repository.create(session);

    await expect(repository.update(session)).resolves.toMatchSnapshot();
  });

  test("should find", async () => {
    await repository.create(session);

    await expect(repository.find({ accountId: "be3a62d1-24a0-401c-96dd-3aff95356811" })).resolves.toMatchSnapshot();
  });

  test("should find many", async () => {
    await repository.create(session);
    await repository.create(
      new Session({
        accountId: "be3a62d1-24a0-401c-96dd-3aff95356811",
        authorization: {
          codeChallenge: "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          codeMethod: "sha256",
          deviceChallenge: "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          email: "email@lindorm.io",
          id: "be3a62d1-24a0-401c-96dd-3aff95356811",
          otpCode: "otpCode",
          redirectUri: "https://lindorm.io/",
          responseType: [ResponseType.REFRESH, ResponseType.ACCESS, ResponseType.IDENTITY].join(" "),
        },
        clientId: "be3a62d1-24a0-401c-96dd-3aff95356811",
        expires: new Date(),
        grantType: GrantType.EMAIL_LINK,
        scope: [Scope.DEFAULT],
      }),
    );

    await expect(repository.findMany({ accountId: "be3a62d1-24a0-401c-96dd-3aff95356811" })).resolves.toMatchSnapshot();
  });

  test("should remove", async () => {
    await repository.create(session);

    await expect(repository.remove(session)).resolves.toBe(undefined);
    await expect(repository.find({ id: session.id })).rejects.toStrictEqual(expect.any(RepositoryEntityNotFoundError));
  });
});
