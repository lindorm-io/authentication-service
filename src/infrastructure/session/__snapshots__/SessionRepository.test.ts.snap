// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionRepository should create 1`] = `
Session {
  "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_agent": Object {
    "browser": "browser",
    "geoIp": Object {
      "geoIp": "geoIp",
    },
    "os": "os",
    "platform": "platform",
    "source": "source",
    "version": "version",
  },
  "_authenticated": false,
  "_authorization": Object {
    "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
    "codeMethod": "sha256",
    "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
    "email": "email@lindorm.io",
    "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "otpCode": "otpCode",
    "redirectUri": "https://lindorm.io/",
    "responseType": "code token id_token",
  },
  "_clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_created": 2020-01-01T07:00:00.000Z,
  "_deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "session_created",
      "payload": Object {
        "accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "agent": Object {
          "browser": "browser",
          "geoIp": Object {
            "geoIp": "geoIp",
          },
          "os": "os",
          "platform": "platform",
          "source": "source",
          "version": "version",
        },
        "authenticated": false,
        "authorization": Object {
          "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          "codeMethod": "sha256",
          "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          "email": "email@lindorm.io",
          "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "otpCode": "otpCode",
          "redirectUri": "https://lindorm.io/",
          "responseType": "code token id_token",
        },
        "clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "expires": 2020-01-01T07:00:00.000Z,
        "grantType": "email_otp",
        "refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "scope": Array [
          "default",
        ],
      },
    },
  ],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_grantType": "email_otp",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_scope": Array [
    "default",
  ],
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`SessionRepository should find 1`] = `
Session {
  "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_agent": Object {
    "browser": "browser",
    "geoIp": Object {
      "geoIp": "geoIp",
    },
    "os": "os",
    "platform": "platform",
    "source": "source",
    "version": "version",
  },
  "_authenticated": false,
  "_authorization": Object {
    "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
    "codeMethod": "sha256",
    "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
    "email": "email@lindorm.io",
    "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "otpCode": "otpCode",
    "redirectUri": "https://lindorm.io/",
    "responseType": "code token id_token",
  },
  "_clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_created": 2020-01-01T07:00:00.000Z,
  "_deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "session_created",
      "payload": Object {
        "accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "agent": Object {
          "browser": "browser",
          "geoIp": Object {
            "geoIp": "geoIp",
          },
          "os": "os",
          "platform": "platform",
          "source": "source",
          "version": "version",
        },
        "authenticated": false,
        "authorization": Object {
          "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          "codeMethod": "sha256",
          "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          "email": "email@lindorm.io",
          "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "otpCode": "otpCode",
          "redirectUri": "https://lindorm.io/",
          "responseType": "code token id_token",
        },
        "clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "expires": 2020-01-01T07:00:00.000Z,
        "grantType": "email_otp",
        "refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "scope": Array [
          "default",
        ],
      },
    },
  ],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_grantType": "email_otp",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_scope": Array [
    "default",
  ],
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`SessionRepository should find many 1`] = `
Array [
  Session {
    "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_agent": Object {
      "browser": "browser",
      "geoIp": Object {
        "geoIp": "geoIp",
      },
      "os": "os",
      "platform": "platform",
      "source": "source",
      "version": "version",
    },
    "_authenticated": false,
    "_authorization": Object {
      "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
      "codeMethod": "sha256",
      "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
      "email": "email@lindorm.io",
      "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
      "otpCode": "otpCode",
      "redirectUri": "https://lindorm.io/",
      "responseType": "code token id_token",
    },
    "_clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_created": 2020-01-01T07:00:00.000Z,
    "_deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_events": Array [
      Object {
        "date": 2020-01-01T07:00:00.000Z,
        "name": "session_created",
        "payload": Object {
          "accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "agent": Object {
            "browser": "browser",
            "geoIp": Object {
              "geoIp": "geoIp",
            },
            "os": "os",
            "platform": "platform",
            "source": "source",
            "version": "version",
          },
          "authenticated": false,
          "authorization": Object {
            "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
            "codeMethod": "sha256",
            "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
            "email": "email@lindorm.io",
            "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
            "otpCode": "otpCode",
            "redirectUri": "https://lindorm.io/",
            "responseType": "code token id_token",
          },
          "clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "expires": 2020-01-01T07:00:00.000Z,
          "grantType": "email_otp",
          "refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "scope": Array [
            "default",
          ],
        },
      },
    ],
    "_expires": 2020-01-01T07:00:00.000Z,
    "_grantType": "email_otp",
    "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_scope": Array [
      "default",
    ],
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
  },
  Session {
    "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_agent": Object {
      "browser": null,
      "geoIp": null,
      "os": null,
      "platform": null,
      "source": null,
      "version": null,
    },
    "_authenticated": false,
    "_authorization": Object {
      "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
      "codeMethod": "sha256",
      "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
      "email": "email@lindorm.io",
      "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
      "otpCode": "otpCode",
      "redirectUri": "https://lindorm.io/",
      "responseType": "code token id_token",
    },
    "_clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_created": 2020-01-01T07:00:00.000Z,
    "_deviceId": null,
    "_events": Array [
      Object {
        "date": 2020-01-01T07:00:00.000Z,
        "name": "session_created",
        "payload": Object {
          "accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "agent": Object {
            "browser": null,
            "geoIp": null,
            "os": null,
            "platform": null,
            "source": null,
            "version": null,
          },
          "authenticated": false,
          "authorization": Object {
            "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
            "codeMethod": "sha256",
            "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
            "email": "email@lindorm.io",
            "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
            "otpCode": "otpCode",
            "redirectUri": "https://lindorm.io/",
            "responseType": "code token id_token",
          },
          "clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "deviceId": null,
          "expires": 2020-01-01T07:00:00.000Z,
          "grantType": "email_link",
          "refreshId": null,
          "scope": Array [
            "default",
          ],
        },
      },
    ],
    "_expires": 2020-01-01T07:00:00.000Z,
    "_grantType": "email_link",
    "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_refreshId": null,
    "_scope": Array [
      "default",
    ],
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
  },
]
`;

exports[`SessionRepository should update 1`] = `
Session {
  "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_agent": Object {
    "browser": "browser",
    "geoIp": Object {
      "geoIp": "geoIp",
    },
    "os": "os",
    "platform": "platform",
    "source": "source",
    "version": "version",
  },
  "_authenticated": false,
  "_authorization": Object {
    "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
    "codeMethod": "sha256",
    "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
    "email": "email@lindorm.io",
    "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "otpCode": "otpCode",
    "redirectUri": "https://lindorm.io/",
    "responseType": "code token id_token",
  },
  "_clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_created": 2020-01-01T07:00:00.000Z,
  "_deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "session_created",
      "payload": Object {
        "accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "agent": Object {
          "browser": "browser",
          "geoIp": Object {
            "geoIp": "geoIp",
          },
          "os": "os",
          "platform": "platform",
          "source": "source",
          "version": "version",
        },
        "authenticated": false,
        "authorization": Object {
          "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          "codeMethod": "sha256",
          "deviceChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
          "email": "email@lindorm.io",
          "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "otpCode": "otpCode",
          "redirectUri": "https://lindorm.io/",
          "responseType": "code token id_token",
        },
        "clientId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "deviceId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "expires": 2020-01-01T07:00:00.000Z,
        "grantType": "email_otp",
        "refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
        "scope": Array [
          "default",
        ],
      },
    },
  ],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_grantType": "email_otp",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_refreshId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_scope": Array [
    "default",
  ],
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 1,
}
`;
